File: examples/scalar_udf/java/app/src/main/java/scalar_udf/FahrenheitToCelsius.java
/**
 * Copyright (c) 2026 Jeffrey Jonathan Jennings
 *
 * @author Jeffrey Jonathan Jennings (J3)
 *
 *
 */
package scalar_udf;

import org.apache.flink.table.functions.ScalarFunction;

/**
 * A Flink scalar user-defined function (UDF) that converts a temperature value
 * from degrees Fahrenheit to degrees Celsius.
 *
 * <p>Being a scalar UDF, it is invoked once per input row and returns exactly
 * one value per invocation. The function is stateless and deterministic: the
 * same input will always produce the same output, which allows Flink's planner
 * to safely cache, reorder, and replay calls during failure recovery.
 *
 * <h3>SQL invocation</h3>
 * <pre>{@code
 * CREATE FUNCTION FahrenheitToCelsius
 *     AS 'scalar_udf.FahrenheitToCelsius'
 *     LANGUAGE JAVA;
 *
 * SELECT temperature_in_fahrenheit,
 *        FahrenheitToCelsius(temperature_in_fahrenheit) AS temperature_in_celsius
 *   FROM readings;
 * }</pre>
 */
public class FahrenheitToCelsius extends ScalarFunction {

    /**
     * Converts a Fahrenheit temperature to its Celsius equivalent using the
     * formula {@code C = (F - 32) * 5/9}.
     *
     * <p>Flink discovers scalar UDF logic by reflecting on public {@code eval}
     * methods, so the method name and visibility must remain as declared.
     *
     * @param fahrenheit the temperature in degrees Fahrenheit; may be {@code null}
     *                to represent a SQL {@code NULL} value
     * @return the temperature in degrees Fahrenheit, or {@code null} if the
     *         input was {@code null} (preserving SQL {@code NULL} semantics)
     */
    public Double eval(Double fahrenheit) {
        if (fahrenheit == null)
            return null;
        return (fahrenheit - 32) * 5.0 / 9.0;
    }
}

File: examples/scalar_udf/python/src/fahrenheit_to_celsius.py
"""
Copyright (c) 2026 Jeffrey Jonathan Jennings

Author: Jeffrey Jonathan Jennings (J3)
"""
from typing import Optional

from pyflink.table.udf import ScalarFunction, udf
from pyflink.table import DataTypes


class FahrenheitToCelsius(ScalarFunction):
    """
    A Flink scalar user-defined function (UDF) that converts a temperature value
    from degrees Fahrenheit to degrees Celsius.

    Being a scalar UDF, it is invoked once per input row and returns exactly one
    value per invocation. The function is stateless and deterministic: the same
    input always produces the same output, which allows Flink's planner to
    safely cache, reorder, and replay calls during failure recovery.

    SQL invocation::

        t_env.create_temporary_system_function(
            "FahrenheitToCelsius",
            fahrenheit_to_celsius,
        )

        t_env.execute_sql('''
            SELECT temperature_in_fahrenheit,
                   FahrenheitToCelsius(temperature_in_fahrenheit) AS temperature_in_celsius
              FROM readings
        ''').print()
    """

    def eval(self, fahrenheit: Optional[float]) -> Optional[float]:
        """
        Convert a Fahrenheit temperature to its Celsius equivalent using the
        formula ``C = (F - 32) * 5/9``.

        PyFlink discovers scalar UDF logic by looking for an ``eval`` method on
        the ``ScalarFunction`` subclass, so the method name must remain as
        declared.

        :param fahrenheit: temperature in degrees Fahrenheit; may be ``None``
            to represent a SQL ``NULL`` value.
        :return: temperature in degrees Celsius, or ``None`` if the input was
            ``None`` (preserving SQL ``NULL`` semantics).
        """
        if fahrenheit is None:
            return None
        return (fahrenheit - 32.0) * 5.0 / 9.0


fahrenheit_to_celsius = udf(
    FahrenheitToCelsius(),
    input_types=[DataTypes.DOUBLE()],
    result_type=DataTypes.DOUBLE(),
)
